File: Export/Natives/Name/FieldFileNamePartProvider.cs
using System;
using System.Collections.Concurrent;
using System.Linq;
using kCura.WinEDDS.Core.Exceptions;
using kCura.WinEDDS.Core.Model;
using kCura.WinEDDS.Core.Model.Export.Process;
using kCura.WinEDDS.Exporters;
using kCura.WinEDDS.Helpers;

namespace kCura.WinEDDS.Core.Export.Natives.Name
{
	public class FieldFileNamePartProvider : FileNamePartProvider<FieldDescriptorPart>
	{
		private readonly ConcurrentDictionary<int, ViewFieldInfo> _cache = new ConcurrentDictionary<int, ViewFieldInfo>();
		
		public override string GetPartName(FieldDescriptorPart descriptorDescriptorPart, ObjectExportInfo exportObject)
		{
			var extExportObject = exportObject as ExtendedObjectExportInfo;

			ViewFieldInfo viewFieldInfo = GetViewField(descriptorDescriptorPart, extExportObject);

			string fieldValue = FieldValueHelper.ConvertToString(extExportObject.GetFieldValue(viewFieldInfo.AvfColumnName), viewFieldInfo, ' ');

			return kCura.Utility.File.Instance.ConvertIllegalCharactersInFilename(fieldValue);
		}

		private ViewFieldInfo GetViewField(FieldDescriptorPart descriptorDescriptorPart, ExtendedObjectExportInfo exportObject)
		{
			if (!_cache.ContainsKey(descriptorDescriptorPart.Value))
			{
				ViewFieldInfo foundViewField = exportObject.SelectedNativeFileNameViewFields.ToList()
					.Find(item => item.AvfId == descriptorDescriptorPart.Value);

				if (foundViewField == null)
				{
					throw new ObjectNotFoundException($"Can not find field id: {descriptorDescriptorPart.Value} in selection list!");
				}
				_cache[descriptorDescriptorPart.Value] = foundViewField;
			}
			return _cache[descriptorDescriptorPart.Value];
		}
	}

}

File: FileNaming/CustomFileNaming/FieldFileNamePartProvider.vb
Imports System.Collections.Concurrent
Imports FileNaming.CustomFileNaming
Imports kCura.WinEDDS.Exporters
Imports kCura.WinEDDS.Helpers

Namespace kCura.WinEDDS.FileNaming.CustomFileNaming
	Public Class FieldFileNamePartProvider
		Inherits FileNamePartProvider(Of FieldDescriptorPart)

		Private ReadOnly _cache As ConcurrentDictionary(Of Integer, ViewFieldInfo) = New ConcurrentDictionary(Of Integer, ViewFieldInfo)()

		Public Overrides Function GetPartName(descriptorPart As FieldDescriptorPart, exportObject As ObjectExportInfo) As String
			Dim extExportObject As ExtendedObjectExportInfo = TryCast(exportObject, ExtendedObjectExportInfo)
			Dim viewFieldInfo As ViewFieldInfo = GetViewField(descriptorPart, extExportObject)
			Dim fieldValueText As String = ConvertToString(extExportObject.GetFieldValue(viewFieldInfo.AvfColumnName), viewFieldInfo, " "c)
			Dim fieldValue As String = ""

			Select Case viewFieldInfo.FieldType
			    Case Relativity.FieldTypeHelper.FieldType.Boolean
				    fieldValue = GetYesNoFieldStatus(viewFieldInfo, fieldValueText)
				Case Relativity.FieldTypeHelper.FieldType.Varchar
					fieldValue = CleanUpFieldValueFromObjectTags(fieldValueText)
				Case Else
					fieldValue = fieldValueText
			End Select


			Return kCura.Utility.File.Instance.ConvertIllegalCharactersInFilename(fieldValue)
		End Function

		Private Function GetYesNoFieldStatus(viewFieldInfo As ViewFieldInfo, fieldValue As String) As String
			Dim retVal As String = ""
			If fieldValue = "True"
				retVal = viewFieldInfo.DisplayName
			Else
				retVal = ""
			End If
			Return retVal
		End Function

		Private Function CleanUpFieldValueFromObjectTags(fieldValue As String) As String
			Return fieldValue.Replace("<object/>", "")
		End Function

		Private Function GetViewField(descriptorPart As FieldDescriptorPart, exportObject As ExtendedObjectExportInfo) As ViewFieldInfo
			If Not _cache.ContainsKey(descriptorPart.Value) Then
				Dim foundViewField As ViewFieldInfo = exportObject.SelectedNativeFileNameViewFields.ToList().
						Find(Function(item) item.FieldArtifactId = descriptorPart.Value)
				If (foundViewField Is Nothing) Then
					Throw New ArgumentOutOfRangeException($"Can not find field id: {descriptorPart.Value} in selection list!")
				End If
				_cache(descriptorPart.Value) = foundViewField
			End If
			Return _cache(descriptorPart.Value)
		End Function
	End Class
End Namespace
